sim/firesim-lib/src/main/scala/bridges/TerminationBridge.scala

```diff
@@ -0,0 +1,136 @@
+
+package firesim.bridges
+
+import chisel3._
+import chisel3.util._
+import freechips.rocketchip.config.Parameters
+import midas.widgets._
+import midas.targetutils._
+import freechips.rocketchip.util.{DecoupledHelper}
+
+/**
+  * Defines a condition under which the simulator should halt. 
+  * @param Message A string printed by the driver to indicate why the simulator is terminating.
+  * @param isError When true, instructs the driver to return a non-zero value when exiting under this condition.
+**/
+case class TerminationCondition(isErr: Boolean, message : String)
+
+//A seq of termination conditions one for each concurrently running program
+case class TerminationBridgeParams(
+  codeMessage: Seq[TerminationCondition]
+)                
+
+class TerminationBridgeTargetIO(params: TerminationBridgeParams) extends Bundle {
+  val clock = Input(Clock())
+  val errCode = Input(UInt((log2Ceil((params.codeMessage).size)).W))
+  val valid = Input(Bool()) 
+}
+
+class TerminationBridgeHostIO(params: TerminationBridgeParams)
+                              (private val targetIO : TerminationBridgeTargetIO = new TerminationBridgeTargetIO(params)) 
+                               extends Bundle with ChannelizedHostPortIO{
+  def targetClockRef = targetIO.clock 
+  //It is just inputs from the target to indicate test completion
+  //The HostPort annotation would not work with aggregates so creating individual channels.
+  val errCodeH = InputChannel(targetIO.errCode)
+  val validH = InputChannel(targetIO.valid)
+}
+
+class TerminationBridge(params: TerminationBridgeParams) extends BlackBox
+    with Bridge[TerminationBridgeHostIO, TerminationBridgeModule] {
+  val io = IO(new TerminationBridgeTargetIO(params))
+  val bridgeIO = new TerminationBridgeHostIO(params)(io)
+
+  val constructorArg = Some(params)
+
+  generateAnnotations()
+}
+
+object TerminationBridge {
+
+  private def annotateTerminationBridge(clock: Clock, lReset: Reset, finish: Seq[Bool], errCode: Seq[TerminationCondition], params: TerminationBridgeParams): TerminationBridge = {
+    val finishAll = finish.reduce(_||_)
+    val errMessageID = PriorityEncoder(finish)
+    //val gReset = WireInit(false.B)
+    //GlobalResetCondition.sink(gReset)
+    val terminationBridgeTarget = Module(new TerminationBridge(params))
+    terminationBridgeTarget.io.errCode := errMessageID
+    terminationBridgeTarget.io.valid := finishAll && !lReset.asBool
+    terminationBridgeTarget.io.clock := clock 
+    terminationBridgeTarget  
+  }
+
+  def apply(clock: Clock, lReset: Reset, finish: Seq[Bool], errCode: Seq[TerminationCondition], params: TerminationBridgeParams): TerminationBridge = {
```

Comment:
Just some naming suggestions on this interface: 
Maybe lReset -> reset, or expand it to "localReset". 

finish -> conditionBools
errCode -> conditionInfo

Also add `ScalaDoc` for both these apply methods, since i htink they are pretty close to being ready to ship.

Reply:
Done.

Reply:
Didn't extensively test it. So removed.